File: water/src/broker/rpc/RpcMessageHeaders.ts
import { getOrThrow } from "../../util/internal/util.js";
import type { MessageId } from "@";
import { BrokerConnection, BrokerMessageHeaders, RpcStatus } from "@";

const RPC_HEADER_NAMES = {
	IN_REPLY_TO: "rpc-in-reply-to",
	STATUS: "rpc-response-status",
	IS_EXCEPTION: "rpc-is-exception",
	IS_UPDATE: "rpc-is-update",
};

export class RpcMessageHeaders extends BrokerMessageHeaders {
	public get inReplyTo(): MessageId {
		return getOrThrow(this.headers, RPC_HEADER_NAMES.IN_REPLY_TO);
	}
	public get status(): RpcStatus {
		return new RpcStatus(safeToInt(getOrThrow(this.headers, RPC_HEADER_NAMES.STATUS)));
	}
	public get isException(): boolean {
		return this.headers[RPC_HEADER_NAMES.IS_EXCEPTION]?.toLowerCase() === "true";
	}
	public get isUpdate(): boolean {
		return this.headers[RPC_HEADER_NAMES.IS_UPDATE]?.toLowerCase() === "true";
	}

	public constructor(base: Record<string, string> | BrokerMessageHeaders);
	public constructor(
		connection: BrokerConnection,
		targetServices: Set<string>,
		targetInstances: Set<string>,
		inReplyTo: MessageId,
		status: RpcStatus,
		isException: boolean,
		isUpdate: boolean,
	);
	public constructor(
		baseOrConnection: Record<string, string> | BrokerMessageHeaders | BrokerConnection,
		targetServices?: Set<string>,
		targetInstances?: Set<string>,
		inReplyTo?: MessageId,
		status?: RpcStatus,
		isException?: boolean,
		isUpdate?: boolean,
	) {
		if (baseOrConnection instanceof BrokerMessageHeaders) {
			super(baseOrConnection.headers);
		} else if (baseOrConnection instanceof BrokerConnection) {
			const connection = baseOrConnection;
			super(
				BrokerMessageHeaders.createHeadersMap(
					connection.serviceName,
					connection.instanceId,
					targetServices!,
					targetInstances!,
					null,
					{
						[RPC_HEADER_NAMES.IN_REPLY_TO]: inReplyTo!,
						[RPC_HEADER_NAMES.STATUS]: status!.code.toString(),
						[RPC_HEADER_NAMES.IS_EXCEPTION]: isException!.toString(),
						[RPC_HEADER_NAMES.IS_UPDATE]: isUpdate!.toString(),
					},
				),
			);
		} else {
			super(baseOrConnection);
		}
	}
}

function safeToInt(value: string): number {
	const parsed = parseInt(value, 10);
	if (isNaN(parsed)) {
		throw new Error(`Failed to parse integer from value: ${value}`);
	}
	return parsed;
}

File: latte/src/main/java/gg/beemo/latte/broker/rpc/RpcMessageHeaders.kt
package gg.beemo.latte.broker.rpc

import gg.beemo.latte.broker.BrokerConnection
import gg.beemo.latte.broker.BrokerMessageHeaders
import gg.beemo.latte.broker.MessageId
import gg.beemo.latte.broker.getOrThrow

class RpcMessageHeaders(headers: Map<String, String>) : BrokerMessageHeaders(headers) {

    val inReplyTo: MessageId by lazy {
        headers.getOrThrow(HEADER_IN_REPLY_TO)
    }
    val status: RpcStatus by lazy {
        RpcStatus(headers.getOrDefault(HEADER_STATUS, "999_999").toInt())
    }
    val isException: Boolean by lazy {
        headers.getOrDefault(HEADER_IS_EXCEPTION, "false").toBoolean()
    }
    val isUpdate: Boolean by lazy {
        headers.getOrDefault(HEADER_IS_UPDATE, "false").toBoolean()
    }

    constructor(base: BrokerMessageHeaders) : this(base.headers)

    constructor(
        connection: BrokerConnection,
        targetServices: Set<String>,
        targetInstances: Set<String>,
        inReplyTo: MessageId,
        status: RpcStatus,
        isException: Boolean,
        isUpdate: Boolean,
    ) : this(
        createHeadersMap(
            connection.serviceName,
            connection.instanceId,
            targetServices,
            targetInstances,
            null,
            extra =
                mapOf(
                    HEADER_IN_REPLY_TO to inReplyTo,
                    HEADER_STATUS to status.code.toString(),
                    HEADER_IS_EXCEPTION to isException.toString(),
                    HEADER_IS_UPDATE to isUpdate.toString(),
                ),
        ),
    )

    companion object {
        private const val HEADER_IN_REPLY_TO = "rpc-in-reply-to"
        private const val HEADER_STATUS = "rpc-response-status"
        private const val HEADER_IS_EXCEPTION = "rpc-is-exception"
        private const val HEADER_IS_UPDATE = "rpc-is-update"
    }
}
